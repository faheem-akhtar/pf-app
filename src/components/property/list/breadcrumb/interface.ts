export interface BreadcrumbInterface {
  /**
   * Name
   * @example Apartments
   */
  name: string;

  /**
   * Routing parameters
   * @example /en/search?t=1&c=1
   */
  link: string;

  /**
   * Total count for the link
   * @example 4881
   */
  count: number;
}
