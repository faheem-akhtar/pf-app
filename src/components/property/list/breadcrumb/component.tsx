import { useRouter } from 'next/router';
import { FunctionComponent, useContext } from 'react';

import { FiltersContext } from 'components/filters/context';

import { PropertyListBreadcrumbComponentPropsInterface } from './component-props.interface';
import { breadcrumbMapQueryToFilters } from './map-query-to-filters';
import { PropertyListBreadcrumbTemplate } from './template';
import { PropertyListBreadcrumbTemplatePropsInterface } from './template-props.interface';

export const PropertyListBreadcrumbComponent: FunctionComponent<PropertyListBreadcrumbComponentPropsInterface> = (
  props
) => {
  const locale = useRouter().locale as string;
  const filtersCtx = useContext(FiltersContext);

  const onClick: PropertyListBreadcrumbTemplatePropsInterface['onClick'] = (event) => {
    event.preventDefault();
    filtersCtx.change((filtersValue) => ({
      ...filtersValue,
      ...breadcrumbMapQueryToFilters(locale, props.link),
    }));
  };

  return <PropertyListBreadcrumbTemplate {...props} onClick={onClick} />;
};
