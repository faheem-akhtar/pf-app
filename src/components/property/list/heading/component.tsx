import { WrapperTemplate } from 'components/wrapper/template';

import styles from '../property-list.module.scss';

export const PropertyListHeadingComponent = ({ pageTitle }: { pageTitle: string }): JSX.Element => {
  const title = pageTitle.split('-')[0];

  return (
    <WrapperTemplate className={styles.heading}>
      <h1 className={styles.title}>{title}</h1>
    </WrapperTemplate>
  );
};
