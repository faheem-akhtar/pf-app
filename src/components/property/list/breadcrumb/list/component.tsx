import { FunctionComponent } from 'react';

import { IconThinChevronDownTemplate } from 'components/icon';
import { domClassMerge } from 'helpers/dom/class-merge';
import { ButtonComponentTypeEnum } from 'library/button/component-type.enum';
import { ButtonIconPositionEnum } from 'library/button/icon-position.enum';
import { ButtonSizeEnum } from 'library/button/size.enum';
import { ButtonTemplate } from 'library/button/template';

import { PropertyListBreadcrumbComponent } from '../component';
import styles from '../property-list-breadcrumb.module.scss';
import { PropertyListBreadcrumbListComponentPropsInterface } from './component-props.interface';
import { propertyListBreadcrumbListItemsVisibleCount } from './items-visible-count';

export const PropertyListBreadcrumbListComponent: FunctionComponent<PropertyListBreadcrumbListComponentPropsInterface> =
  ({ breadcrumbs, onClickShowMore, expanded, t }) => (
    <section role='section' className={styles.breadcrumb}>
      <ul role='list' className={styles.breadcrumb__list}>
        {breadcrumbs.slice(0, propertyListBreadcrumbListItemsVisibleCount).map((breadcrumb) => (
          <PropertyListBreadcrumbComponent key={breadcrumb.name} {...breadcrumb} />
        ))}

        {breadcrumbs.length > propertyListBreadcrumbListItemsVisibleCount && (
          <li role='listitem'>
            <ButtonTemplate
              className={styles.breadcrumb__button}
              type='button'
              componentType={ButtonComponentTypeEnum.secondaryBlue}
              size={ButtonSizeEnum.small}
              onClick={onClickShowMore}
              icon={{
                component: IconThinChevronDownTemplate,
                position: ButtonIconPositionEnum.right,
                className: domClassMerge(styles.breadcrumb__icon, {
                  [styles['breadcrumb__icon--expanded']]: expanded,
                }),
              }}
            >
              {t(expanded ? 'show-less' : 'show-more')}
            </ButtonTemplate>
          </li>
        )}
      </ul>

      <ul
        className={domClassMerge(styles.breadcrumb__list, styles['breadcrumb__list--more'], {
          [styles['breadcrumb__list--expanded']]: expanded,
        })}
      >
        {breadcrumbs.slice(propertyListBreadcrumbListItemsVisibleCount, breadcrumbs.length).map((breadcrumb) => (
          <PropertyListBreadcrumbComponent
            key={breadcrumb.name}
            class={styles['breadcrumb__item--more']}
            {...breadcrumb}
          />
        ))}
      </ul>
    </section>
  );
