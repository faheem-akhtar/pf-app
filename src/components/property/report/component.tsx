import { MutableRefObject, useContext, useRef, useState } from 'react';

import { apiReportAttachmentsFetcher } from 'api/report/attachments/fetcher';
import { apiReportFetcher } from 'api/report/fetcher';
import { AuthModalComponent } from 'components/auth/modal/component';
import { ModalComponent } from 'components/modal/component';
import { UserContext } from 'context/user/context';
import { functionNoop } from 'helpers/function/noop';
import { GoogleRecaptchaService } from 'services/google/recaptcha.service';
import { TFunctionType } from 'types/t-function/type';

import { PropertyReportFormSubmitPayloadInterface } from './form/submit-payload.interface';
import { PropertyReportTemplate } from './template';

const captchaService = GoogleRecaptchaService();

export const PropertyReportComponent = ({
  openRef,
  propertyId,
  t,
}: {
  openRef: MutableRefObject<() => void>;
  propertyId: string;
  t: TFunctionType;
}): JSX.Element => {
  const [isRequestOnProgress, setIsRequestOnProgress] = useState(false);
  const [isReportSent, setIsReportSent] = useState(false);
  const [requestErrorMessage, setRequestErrorMessage] = useState('');
  const user = useContext(UserContext);

  const closeRef = useRef<() => void>(functionNoop);

  const closeModal = (): void => {
    closeRef.current();
  };

  const closeReportModal = (): void => {
    setIsReportSent(false);
    setIsRequestOnProgress(false);
    setRequestErrorMessage('');
    closeModal();
  };

  const onClickSubmit = async (payload: PropertyReportFormSubmitPayloadInterface): Promise<void> => {
    setIsRequestOnProgress(true);
    await captchaService.execute();
    setRequestErrorMessage('');
    const { attachment, ...attributes } = payload;
    const requests = [
      (): ReturnType<typeof apiReportFetcher> =>
        apiReportFetcher(propertyId, {
          ...attributes,
          email: user?.email || '',
        }),
    ];

    if (attachment) {
      const fileData = new FormData();

      fileData.append('attachment[]', attachment, attachment.name);
      requests.push(() => apiReportAttachmentsFetcher(propertyId, fileData));
    }

    Promise.all(requests.map((request) => request())).then((responses) => {
      setIsRequestOnProgress(false);
      if (responses.every((response) => response.ok)) {
        setIsReportSent(true);
      } else {
        setRequestErrorMessage(t('something-wrong-try-again'));
        captchaService.reset();
      }
    });
  };

  return (
    <ModalComponent
      openRef={openRef}
      closeRef={closeRef}
      overlay
      onOverlayClick={!user ? closeModal : closeReportModal}
    >
      {!user ? (
        <AuthModalComponent
          close={closeModal}
          success={(): void => {
            closeModal();
            openRef.current();
          }}
        />
      ) : (
        <PropertyReportTemplate
          isReportSent={isReportSent}
          onSubmit={onClickSubmit}
          isLoading={isRequestOnProgress}
          errorMessage={requestErrorMessage}
          t={t}
          onClose={closeReportModal}
        />
      )}
    </ModalComponent>
  );
};
