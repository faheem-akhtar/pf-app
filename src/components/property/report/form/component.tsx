import { useContext, useRef, useState } from 'react';

import { ErrorMessageComponent } from 'components/error-message/component';
import { FiltersContext } from 'components/filters/context';
import { formMakeValidator } from 'components/form/make-validator';
import { ReCaptchaComponent } from 'components/re-captcha/component';
import { configPropertyReportAttachmentSizeLimit } from 'config/property/report-attachment-size-limit';
import { configPropertyReportMessageMinCharCount } from 'config/property/report-message-min-char-count';
import { FiltersParametersEnum } from 'enums/filters/parameters.enum';
import { PropertyReportReasonEnum } from 'enums/property/report/reason.enum';
import { PropertyReportUserTypeEnum } from 'enums/property/report/user-type.enum';
import { categoryIdIsSale } from 'helpers/category-id/is-sale';
import { userReportCategory } from 'helpers/user/report-category';
import { validationMinLength } from 'helpers/validation/min-length';
import { validationRequired } from 'helpers/validation/required';
import { ButtonComponentTypeEnum } from 'library/button/component-type.enum';
import { ButtonSizeEnum } from 'library/button/size.enum';
import { ButtonTemplate } from 'library/button/template';
import { InputBaseComponent } from 'library/input/base/component';
import { SelectFieldTemplate } from 'library/select-field/template';

import { propertyReportFormAdditionalUserTypes } from './additional-user-types';
import { PropertyReportFormComponentPropsInterface } from './component-props.interface';
import { PropertyReportFormFieldsEnum } from './fields.enum';
import styles from './property-report-form.module.scss';
import { propertyReportFormReasons } from './reasons';

export const PropertyReportFormComponent = ({
  t,
  onClickSubmit,
  loading,
  errorMessage,
}: PropertyReportFormComponentPropsInterface): JSX.Element => {
  const fileInputRef = useRef<HTMLInputElement>(null);
  const filtersCtx = useContext(FiltersContext);

  const [reasonType, setReasonType] = useState<PropertyReportReasonEnum | ''>('');
  const [userType, setUserType] = useState<PropertyReportUserTypeEnum | ''>('');
  const [attachment, setAttachment] = useState<File>();
  const [message, setMessage] = useState<string>('');

  const [errors, setErrors] = useState<Partial<Record<PropertyReportFormFieldsEnum, string>>>({});
  const [validators] = useState(() => ({
    [PropertyReportFormFieldsEnum.reason]: [validationRequired(t('report-modal/empty-reason'))],
    [PropertyReportFormFieldsEnum.user]: [validationRequired(t('report-modal/empty-user-type'))],
    [PropertyReportFormFieldsEnum.message]: [
      validationRequired(t('validation/empty-message')),
      validationMinLength(t('report-modal/not-valid-message-length'), configPropertyReportMessageMinCharCount),
    ],
  }));

  const validate = formMakeValidator(errors, setErrors, validators);

  const categoryId = filtersCtx.value[FiltersParametersEnum.categoryId];

  return (
    <form
      onSubmit={(e): void => {
        e.preventDefault();
        if (
          !validate({
            [PropertyReportFormFieldsEnum.reason]: reasonType,
            [PropertyReportFormFieldsEnum.user]: userType,
            [PropertyReportFormFieldsEnum.message]: message,
          })
        ) {
          onClickSubmit({
            reason_id: reasonType as PropertyReportReasonEnum,
            reporter_type: userType as PropertyReportUserTypeEnum,
            email: '',
            message,
            attachment,
          });
        }
      }}
    >
      <SelectFieldTemplate
        onChange={(value): void => {
          validate({ [PropertyReportFormFieldsEnum.reason]: value });
          setReasonType(value);
        }}
        value={reasonType}
        error={!!errors[PropertyReportFormFieldsEnum.reason]}
        errorText={errors[PropertyReportFormFieldsEnum.reason]}
        className={styles.field}
        options={[
          {
            label: t('report-modal/select-reason'),
            value: '',
          },
          ...propertyReportFormReasons.map(({ label, value }) => ({
            value,
            label: t(label),
          })),
        ]}
      />
      <SelectFieldTemplate
        onChange={(value): void => {
          validate({ [PropertyReportFormFieldsEnum.user]: value });
          setUserType(value as PropertyReportUserTypeEnum);
        }}
        value={userType}
        error={!!errors[PropertyReportFormFieldsEnum.user]}
        errorText={errors[PropertyReportFormFieldsEnum.user]}
        className={styles.field}
        options={[
          {
            label: t('report-modal/select-user-type'),
            value: '',
          },
          {
            label: t(`user-${categoryIdIsSale(categoryId) ? 'buyer' : 'renter'}`),
            value: userReportCategory(categoryId),
          },
          ...propertyReportFormAdditionalUserTypes.map(({ label, value }) => ({
            value,
            label: t(label),
          })),
        ]}
      />
      {userType === PropertyReportUserTypeEnum.agent && (
        <div className={styles.field}>
          <span className={styles.text}>
            {t('report-modal/max-file-size', { size: configPropertyReportAttachmentSizeLimit })}
          </span>
          <input
            type='file'
            name='attachment'
            ref={fileInputRef}
            onChange={(event): void => {
              const { files } = event.target;
              if (files) {
                setAttachment(files[0]);
              }
            }}
          />
        </div>
      )}
      <div className={styles.messageContainer}>
        <span>{t('report-modal/additional-messages')}</span>
        <InputBaseComponent
          value={message}
          className={styles.message}
          textarea
          onChange={(value): void => {
            validate({ [PropertyReportFormFieldsEnum.message]: value });
            setMessage(value);
          }}
          error={!!errors[PropertyReportFormFieldsEnum.message]}
          errorText={errors[PropertyReportFormFieldsEnum.message]}
        />
      </div>
      {errorMessage && <ErrorMessageComponent message={errorMessage} />}
      <ButtonTemplate
        type='submit'
        loading={loading}
        className={styles.send}
        size={ButtonSizeEnum.small}
        componentType={ButtonComponentTypeEnum.primary}
      >
        {`${t('send')} ${t('cta-report')}`}
      </ButtonTemplate>
      <ReCaptchaComponent />
    </form>
  );
};
