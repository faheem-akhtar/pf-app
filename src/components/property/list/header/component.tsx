import { FunctionComponent, useState } from 'react';

import { WrapperTemplate } from 'components/wrapper/template';
import { useTranslation } from 'helpers/translation/hook';

import { PropertyListBreadcrumbListComponent } from '../breadcrumb/list/component';
import { PropertyListHeaderComponentPropsInterface } from './component-props.interface';
import styles from './property-list-header.module.scss';

export const PropertyListHeaderComponent: FunctionComponent<PropertyListHeaderComponentPropsInterface> = (props) => {
  const { t } = useTranslation();
  const [isExpanded, setIsExpanded] = useState<boolean>(false);

  const pageTitle = props.pageTitle.split('-')[0];

  return (
    <WrapperTemplate className={styles.heading}>
      <>
        <h1 className={styles.title}>{pageTitle}</h1>

        {props.breadcrumbs.length ? (
          <PropertyListBreadcrumbListComponent
            breadcrumbs={props.breadcrumbs}
            onClickShowMore={(): void => setIsExpanded((isExpanded) => !isExpanded)}
            expanded={isExpanded}
            t={t}
          />
        ) : null}
      </>
    </WrapperTemplate>
  );
};
