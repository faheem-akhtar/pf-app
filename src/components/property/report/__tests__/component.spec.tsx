import { act, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { mockModalEnv } from 'mocks/modal-env/mock';
import { mockWindowFetch } from 'mocks/window/fetch.mock';
import { filtersContextPropsStub } from 'stubs/filters/context-props.stub';
import { googleRecaptchaStub } from 'stubs/google/recaptcha.stub';
import { propertyStub } from 'stubs/property/stub';
import { userModelStub } from 'stubs/user/model.stub';

import * as AuthLoginComponentModule from 'components/auth/login/component';
import { AuthSuccessTypeEnum } from 'components/auth/success-type.enum';
import { FiltersContextProvider } from 'components/filters/context-provider';
import { propertySerpObfuscatedGetId } from 'components/property/serp/obfuscated/get/id';
import { PropertySerpObfuscatedType } from 'components/property/serp/obfuscated/type';
import { UserContext } from 'components/user/context';
import { FiltersCategoryIdEnum } from 'enums/filters/category-id.enum';
import { FiltersParametersEnum } from 'enums/filters/parameters.enum';
import { functionSelf } from 'helpers/function/self';
import { GoogleRecaptcha } from 'services/google/recaptcha';
import * as GoogleRecaptchaServiceModule from 'services/google/recaptcha.service';

import { PropertyReportComponent } from '../component';
import { PropertyReportComponentPropsInterface } from '../component-props.interface';

describe('PropertyReportComponent', () => {
  let props: PropertyReportComponentPropsInterface;
  const openRef = { current: jest.fn() };
  const property: PropertySerpObfuscatedType = propertyStub();
  const propertyId: string = propertySerpObfuscatedGetId(property);

  beforeEach(() => {
    mockModalEnv();

    props = {
      openRef,
      propertyId,
      t: functionSelf,
    };
  });

  it('renders without throwing any errors', () => {
    render(
      <UserContext.Provider value={userModelStub()}>
        <FiltersContextProvider {...filtersContextPropsStub()}>
          <PropertyReportComponent {...props} />
        </FiltersContextProvider>
      </UserContext.Provider>
    );
    act(openRef.current);

    expect(screen.getByTestId('property-report')).toMatchSnapshot();
  });

  it('should display report modal content after login is succeed', async () => {
    const { filtersValueFromQuery, filtersData } = filtersContextPropsStub();

    jest
      .spyOn(AuthLoginComponentModule, 'AuthLoginComponent')
      .mockImplementationOnce(({ onSuccess }) => (
        <button
          data-testid='login-btn'
          onClick={(): void => onSuccess(AuthSuccessTypeEnum.signInWithEmail, userModelStub())}
        />
      ));

    const { rerender } = render(
      <UserContext.Provider value={null}>
        <FiltersContextProvider
          filtersValueFromQuery={{
            ...filtersValueFromQuery,
            [FiltersParametersEnum.categoryId]: FiltersCategoryIdEnum.commercialForSale,
          }}
          filtersData={filtersData}
        >
          <PropertyReportComponent {...props} />
        </FiltersContextProvider>
      </UserContext.Provider>
    );
    act(openRef.current);

    userEvent.click(screen.getByTestId('login-btn'));

    rerender(
      <UserContext.Provider value={userModelStub()}>
        <FiltersContextProvider
          filtersValueFromQuery={{
            ...filtersValueFromQuery,
            [FiltersParametersEnum.categoryId]: FiltersCategoryIdEnum.commercialForSale,
          }}
          filtersData={filtersData}
        >
          <PropertyReportComponent {...props} />
        </FiltersContextProvider>
      </UserContext.Provider>
    );

    expect(screen.getByTestId('property-report')).toBeInTheDocument();
    expect(screen.getByRole('option', { name: 'user-buyer' })).toBeInTheDocument();
    expect(screen.queryByRole('option', { name: 'user-renter' })).not.toBeInTheDocument();
  });

  it('should reset the state values when the modal is closed', () => {
    render(
      <UserContext.Provider value={userModelStub()}>
        <FiltersContextProvider {...filtersContextPropsStub()}>
          <PropertyReportComponent {...props} />
        </FiltersContextProvider>
      </UserContext.Provider>
    );
    act(openRef.current);

    expect(screen.getByText('report-modal/title')).toBeInTheDocument();

    userEvent.click(screen.getByRole('button', { name: 'send cta-report' }));

    expect(screen.getByText('report-modal/empty-reason')).toBeInTheDocument();
    expect(screen.getByText('report-modal/empty-user-type')).toBeInTheDocument();
    expect(screen.getByText('validation/empty-message')).toBeInTheDocument();

    userEvent.click(screen.getByRole('button', { name: /close/i }));

    expect(screen.queryByText('report-modal/title')).not.toBeInTheDocument();

    render(
      <UserContext.Provider value={userModelStub()}>
        <FiltersContextProvider {...filtersContextPropsStub()}>
          <PropertyReportComponent {...props} />
        </FiltersContextProvider>
      </UserContext.Provider>
    );
    act(openRef.current);

    expect(screen.queryByText('report-modal/empty-reason')).not.toBeInTheDocument();
    expect(screen.queryByText('report-modal/empty-user-type')).not.toBeInTheDocument();
    expect(screen.queryByText('validation/empty-message')).not.toBeInTheDocument();
  });

  test('if the user type agent', () => {
    render(
      <UserContext.Provider value={userModelStub()}>
        <FiltersContextProvider {...filtersContextPropsStub()}>
          <PropertyReportComponent {...props} />
        </FiltersContextProvider>
      </UserContext.Provider>
    );
    act(openRef.current);

    userEvent.selectOptions(screen.getAllByRole('combobox')[1], screen.getByRole('option', { name: 'user-agent' }));

    expect(screen.getByText('report-modal/max-file-size')).toBeInTheDocument();

    userEvent.selectOptions(screen.getAllByRole('combobox')[1], screen.getByRole('option', { name: 'user-landlord' }));

    expect(screen.queryByText('report-modal/max-file-size')).not.toBeInTheDocument();
  });

  describe('Form Submission', () => {
    let fetchMock: jest.Mock;
    let googleRecaptchaMock: GoogleRecaptcha;

    beforeEach(() => {
      fetchMock = mockWindowFetch();

      googleRecaptchaMock = googleRecaptchaStub();
      jest.spyOn(GoogleRecaptchaServiceModule, 'GoogleRecaptchaService').mockReturnValue(googleRecaptchaMock);

      render(
        <UserContext.Provider value={userModelStub()}>
          <FiltersContextProvider {...filtersContextPropsStub()}>
            <PropertyReportComponent {...props} />
          </FiltersContextProvider>
        </UserContext.Provider>
      );
      act(openRef.current);
    });

    it('should submit the form and display success', async () => {
      userEvent.selectOptions(
        screen.getAllByRole('combobox')[0],
        screen.getByRole('option', { name: 'report/no-response-broker' })
      );
      userEvent.selectOptions(screen.getAllByRole('combobox')[1], screen.getByRole('option', { name: 'user-agent' }));
      userEvent.upload(
        screen.getByTestId('property-report-upload-file'),
        new File(['hello'], 'hello.png', { type: 'image/png' })
      );
      userEvent.type(screen.getByPlaceholderText('report-modal/additional-messages-placeholder'), 'lorem ipsum');
      userEvent.click(screen.getByRole('button', { name: 'send cta-report' }));

      await waitFor(() =>
        expect(fetchMock).toHaveBeenCalledWith(`default-origin/en/api/property/${propertyId}/report`, {
          body: JSON.stringify({
            data: {
              type: 'property_report',
              attributes: {
                reason_id: 3,
                reporter_type: 'agent',
                email: 'test@propertyfinder.ae',
                message: 'lorem ipsum',
              },
            },
          }),
          headers: {
            locale: 'en',
            'content-type': 'application/vnd.api+json',
          },
          method: 'POST',
        })
      );

      expect(fetchMock).toHaveBeenCalledTimes(2);
      expect(screen.getByText('thank-you')).toBeInTheDocument();
      expect(screen.getByText('report-modal/report-is-sent')).toBeInTheDocument();
      expect(googleRecaptchaMock.execute).toHaveBeenCalledTimes(1);
      expect(googleRecaptchaMock.reset).not.toHaveBeenCalled();
    });

    it('should display general error message if the request failed', async () => {
      fetchMock = mockWindowFetch({ ok: false, status: 500 });

      userEvent.selectOptions(
        screen.getAllByRole('combobox')[0],
        screen.getByRole('option', { name: 'report/no-response-broker' })
      );
      userEvent.selectOptions(
        screen.getAllByRole('combobox')[1],
        screen.getByRole('option', { name: 'user-landlord' })
      );
      userEvent.type(screen.getByPlaceholderText('report-modal/additional-messages-placeholder'), 'lorem ipsum');
      userEvent.click(screen.getByRole('button', { name: 'send cta-report' }));

      await screen.findByText('something-wrong-try-again');

      expect(screen.getByText('something-wrong-try-again'));

      expect(googleRecaptchaMock.execute).toHaveBeenCalledTimes(1);
      expect(googleRecaptchaMock.reset).toHaveBeenCalledTimes(1);
    });

    it('should select an option', () => {
      userEvent.selectOptions(
        screen.getAllByRole('combobox')[0],
        screen.getByRole('option', { name: 'report/no-response-broker' })
      );

      expect(
        (screen.getByRole('option', { name: 'report/no-response-broker' }) as HTMLOptionElement).selected
      ).toBeTruthy();

      userEvent.selectOptions(screen.getAllByRole('combobox')[1], screen.getByRole('option', { name: 'user-renter' }));

      expect((screen.getByRole('option', { name: 'user-renter' }) as HTMLOptionElement).selected).toBeTruthy();
    });

    it('should not submit the form and display the error for reason type', () => {
      userEvent.selectOptions(screen.getAllByRole('combobox')[1], screen.getByRole('option', { name: 'user-renter' }));
      userEvent.type(screen.getByPlaceholderText('report-modal/additional-messages-placeholder'), 'lorem ipsum');
      userEvent.click(screen.getByRole('button', { name: 'send cta-report' }));

      expect(screen.getByText('report-modal/empty-reason')).toBeInTheDocument();
    });

    it('should not submit the form and display the error for user type', () => {
      userEvent.selectOptions(
        screen.getAllByRole('combobox')[0],
        screen.getByRole('option', { name: 'report/no-response-broker' })
      );
      userEvent.type(screen.getByPlaceholderText('report-modal/additional-messages-placeholder'), 'lorem ipsum');
      userEvent.click(screen.getByRole('button', { name: 'send cta-report' }));

      expect(screen.getByText('report-modal/empty-user-type')).toBeInTheDocument();
    });

    it('should not submit the form and display the error for message', () => {
      userEvent.selectOptions(
        screen.getAllByRole('combobox')[0],
        screen.getByRole('option', { name: 'report/no-response-broker' })
      );
      userEvent.selectOptions(screen.getAllByRole('combobox')[1], screen.getByRole('option', { name: 'user-renter' }));
      userEvent.click(screen.getByRole('button', { name: 'send cta-report' }));

      expect(screen.getByText('validation/empty-message')).toBeInTheDocument();

      userEvent.type(screen.getByPlaceholderText('report-modal/additional-messages-placeholder'), 'amet');

      expect(screen.getByText('report-modal/not-valid-message-length')).toBeInTheDocument();
    });

    it('should not submit the form and display the error for file', () => {
      userEvent.selectOptions(
        screen.getAllByRole('combobox')[0],
        screen.getByRole('option', { name: 'report/no-response-broker' })
      );
      userEvent.selectOptions(screen.getAllByRole('combobox')[1], screen.getByRole('option', { name: 'user-agent' }));
      userEvent.type(screen.getByPlaceholderText('report-modal/additional-messages-placeholder'), 'amet');
      userEvent.upload(
        screen.getByTestId('property-report-upload-file'),
        new File(['hello'], 'hello.gif', { type: 'image/gif' })
      );
      userEvent.click(screen.getByRole('button', { name: 'send cta-report' }));

      expect(screen.getByText('report-modal/file-type-error')).toBeInTheDocument();
    });
  });
});
