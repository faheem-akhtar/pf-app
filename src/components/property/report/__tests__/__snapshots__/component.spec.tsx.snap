// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyReportComponent renders without throwing any errors 1`] = `
<div
  class="container"
  data-testid="property-report"
>
  <div
    class="header"
  >
    <button
      class="close"
    >
      <svg
        class="closeIcon"
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
        <title>
          Close
        </title>
        <path
          clip-rule="evenodd"
          d="M5.6129 4.2097L5.70711 4.29289L12 10.585L18.2929 4.29289C18.6834 3.90237 19.3166 3.90237 19.7071 4.29289C20.0676 4.65338 20.0953 5.22061 19.7903 5.6129L19.7071 5.70711L13.414 12L19.7071 18.2929C20.0976 18.6834 20.0976 19.3166 19.7071 19.7071C19.3466 20.0676 18.7794 20.0953 18.3871 19.7903L18.2929 19.7071L12 13.414L5.70711 19.7071C5.31658 20.0976 4.68342 20.0976 4.29289 19.7071C3.93241 19.3466 3.90468 18.7794 4.2097 18.3871L4.29289 18.2929L10.585 12L4.29289 5.70711C3.90237 5.31658 3.90237 4.68342 4.29289 4.29289C4.62334 3.96245 5.12751 3.91161 5.5114 4.14038L5.6129 4.2097Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <svg
      class="icon"
      fill="none"
      viewBox="0 0 1792 1792"
    >
      <path
        d="M1024 1375v-190q0-14-9.5-23.5t-22.5-9.5h-192q-13 0-22.5 9.5t-9.5 23.5v190q0 14 9.5 23.5t22.5 9.5h192q13 0 22.5-9.5t9.5-23.5zm-2-374l18-459q0-12-10-19-13-11-24-11h-220q-11 0-24 11-10 7-10 21l17 457q0 10 10 16.5t24 6.5h185q14 0 23.5-6.5t10.5-16.5zm-14-934l768 1408q35 63-2 126-17 29-46.5 46t-63.5 17h-1536q-34 0-63.5-17t-46.5-46q-37-63-2-126l768-1408q17-31 47-49t65-18 65 18 47 49z"
      />
    </svg>
    <span>
      report-modal/title
    </span>
  </div>
  <form>
    <div
      class="container field"
    >
      <div
        class="fieldWrapper"
      >
        <select
          class="field fieldEmpty fieldWithIcon"
        >
          <option
            value="0"
          >
            report-modal/select-reason
          </option>
          <option
            value="1"
          >
            report/property-not-available
          </option>
          <option
            value="2"
          >
            report/inaccurate-price
          </option>
          <option
            value="3"
          >
            report/no-response-broker
          </option>
          <option
            value="4"
          >
            report/no-property-details
          </option>
          <option
            value="5"
          >
            report/poor-quality-irrelevant-photos
          </option>
          <option
            value="6"
          >
            report/poor-description
          </option>
          <option
            value="7"
          >
            report/inaccurate-location
          </option>
          <option
            value="8"
          >
            report/inaccurate-bedrooms
          </option>
          <option
            value="9"
          >
            report/incorrect-property-type
          </option>
        </select>
        <svg
          class="icon"
          viewBox="4.5 8 16 8"
        >
          <path
            clip-rule="evenodd"
            d="M4.64645 8.14645C4.47288 8.32001 4.4536 8.58944 4.58859 8.78431L4.64645 8.85355L11.6464 15.8536C11.82 16.0271 12.0894 16.0464 12.2843 15.9114L12.3536 15.8536L19.3536 8.85355C19.5488 8.65829 19.5488 8.34171 19.3536 8.14645C19.18 7.97288 18.9106 7.9536 18.7157 8.08859L18.6464 8.14645L12 14.793L5.35355 8.14645C5.17999 7.97288 4.91056 7.9536 4.71569 8.08859L4.64645 8.14645Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      
    </div>
    <div
      class="container field"
    >
      <div
        class="fieldWrapper"
      >
        <select
          class="field fieldEmpty fieldWithIcon"
        >
          <option
            value="0"
          >
            report-modal/select-user-type
          </option>
          <option
            value="1"
          >
            user-renter
          </option>
          <option
            value="2"
          >
            user-landlord
          </option>
          <option
            value="3"
          >
            user-agent
          </option>
        </select>
        <svg
          class="icon"
          viewBox="4.5 8 16 8"
        >
          <path
            clip-rule="evenodd"
            d="M4.64645 8.14645C4.47288 8.32001 4.4536 8.58944 4.58859 8.78431L4.64645 8.85355L11.6464 15.8536C11.82 16.0271 12.0894 16.0464 12.2843 15.9114L12.3536 15.8536L19.3536 8.85355C19.5488 8.65829 19.5488 8.34171 19.3536 8.14645C19.18 7.97288 18.9106 7.9536 18.7157 8.08859L18.6464 8.14645L12 14.793L5.35355 8.14645C5.17999 7.97288 4.91056 7.9536 4.71569 8.08859L4.64645 8.14645Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      
    </div>
    <div
      class="messageContainer"
    >
      <span>
        report-modal/additional-messages
      </span>
      <div
        class="container message container--textarea"
      >
        <div
          class="content"
        >
          <textarea
            aria-errormessage=""
            aria-invalid="false"
            class="input input--textarea"
            placeholder="report-modal/additional-messages-placeholder"
            type="text"
          />
        </div>
      </div>
      
    </div>
    
    <button
      class="button primary regular fullWidth send"
    >
      send cta-report
    </button>
    <div
      id="recaptcha"
    />
  </form>
</div>
`;
